core/src/main/java/com/linecorp/armeria/server/ServerListenerBuilder.java

```diff
@@ -0,0 +1,309 @@
+     * */
+    private final List<Consumer<? super Server>> serverStoppedCallbacks = new ArrayList<>();
+
+    private static class CallbackServerListener implements ServerListener {
+        /**
+         * {@link Consumer}s invoked when the {@link Server} is starting.
+         * */
+        private final List<Consumer<? super Server>> serverStartingCallbacks;
+
+        /**
+         * {@link Consumer}s invoked when the {@link Server} is started.
+         * */
+        private final List<Consumer<? super Server>> serverStartedCallbacks;
+
+        /**
+         * {@link Consumer}s invoked when the {@link Server} is stopping.
+         * */
+        private final List<Consumer<? super Server>> serverStoppingCallbacks;
+
+        /**
+         * {@link Consumer}s invoked when the {@link Server} is stopped.
+         * */
+        private final List<Consumer<? super Server>> serverStoppedCallbacks;
+
+        CallbackServerListener(List<Consumer<? super Server>> serverStartingCallbacks,
+                               List<Consumer<? super Server>> serverStartedCallbacks,
+                               List<Consumer<? super Server>> serverStoppingCallbacks,
+                               List<Consumer<? super Server>> serverStoppedCallbacks) {
+            this.serverStartingCallbacks = ImmutableList.copyOf(serverStartingCallbacks);
+            this.serverStartedCallbacks = ImmutableList.copyOf(serverStartedCallbacks);
+            this.serverStoppingCallbacks = ImmutableList.copyOf(serverStoppingCallbacks);
+            this.serverStoppedCallbacks = ImmutableList.copyOf(serverStoppedCallbacks);
+        }
+
+        @Override
+        public void serverStarting(Server server) {
+            for (Consumer<? super Server> serverStartingCallback : serverStartingCallbacks) {
+                serverStartingCallback.accept(server);
+            }
+        }
+
+        @Override
+        public void serverStarted(Server server) {
+            for (Consumer<? super Server> serverStartedCallback : serverStartedCallbacks) {
+                serverStartedCallback.accept(server);
+            }
+        }
+
+        @Override
+        public void serverStopping(Server server) {
+            for (Consumer<? super Server> serverStoppingCallback : serverStoppingCallbacks) {
+                serverStoppingCallback.accept(server);
+            }
+        }
+
+        @Override
+        public void serverStopped(Server server) {
+            for (Consumer<? super Server> serverStoppedCallback : serverStoppedCallbacks) {
+                serverStoppedCallback.accept(server);
+            }
+        }
+    }
+
+    /**
+     * Add {@link Runnable} invoked when the {@link Server} is starting.
+     * (see: {@link ServerListener#serverStarting(Server)})
+     */
+    public ServerListenerBuilder addStartingCallback(Runnable runnable) {
+        requireNonNull(runnable, "runnable");
+        serverStartingCallbacks.add((Server server) -> runnable.run());
```

Comment:
Do you need the type parameter? (not sure, maybe you do since it's a wildcard).

Also consider naming the variable `unused`